src/backend/distributed/utils/node_metadata.c

```diff
@@ -0,0 +1,722 @@
+								   workerNodeLine),
+							errhint("Lines in the worker node file must contain a valid "
+									"node name and, optionally, a positive port number. "
+									"Comments begin with a '#' character and extend to "
+									"the end of their line.")));
+		}
+
+		/* allocate worker node structure and set fields */
+		workerNode = (WorkerNode *) palloc0(sizeof(WorkerNode));
+
+		strlcpy(workerNode->workerName, nodeName, WORKER_LENGTH);
+		workerNode->workerPort = nodePort;
+
+		workerNodeList = lappend(workerNodeList, workerNode);
+	}
+
+	FreeFile(workerFileStream);
+	free(workerFilePath);
+
+	/* rename the file, marking that it is not used anymore */
+	appendStringInfo(renamedWorkerFilePath, "%s", workerFilePath);
+	appendStringInfo(renamedWorkerFilePath, ".obsolete");
+	rename(workerFilePath, renamedWorkerFilePath->data);
+
+	return workerNodeList;
+}
+
+
+/*
+ * TupleToWorkerNode takes in a heap tuple from pg_dist_node, and
+ * converts this tuple to an equivalent struct in memory. The function assumes
+ * the caller already has locks on the tuple, and doesn't perform any locking.
+ */
+static WorkerNode *
+TupleToWorkerNode(TupleDesc tupleDescriptor, HeapTuple heapTuple)
+{
+	WorkerNode *workerNode = NULL;
+	bool isNull = false;
+
+	Datum nodeId = heap_getattr(heapTuple, Anum_pg_dist_node_nodeid,
+								tupleDescriptor, &isNull);
+	Datum groupId = heap_getattr(heapTuple, Anum_pg_dist_node_groupid,
+								 tupleDescriptor, &isNull);
+	Datum nodeName = heap_getattr(heapTuple, Anum_pg_dist_node_nodename,
+								  tupleDescriptor, &isNull);
+	Datum nodePort = heap_getattr(heapTuple, Anum_pg_dist_node_nodeport,
+								  tupleDescriptor, &isNull);
+	Datum nodeRack = heap_getattr(heapTuple, Anum_pg_dist_node_noderack,
+								  tupleDescriptor, &isNull);
+
+	Assert(!HeapTupleHasNulls(heapTuple));
+
+	workerNode = (WorkerNode *) palloc0(sizeof(WorkerNode));
+	workerNode->nodeId = DatumGetUInt32(nodeId);
+	workerNode->workerPort = DatumGetUInt32(nodePort);
+	workerNode->groupId = DatumGetUInt32(groupId);
+	strlcpy(workerNode->workerName, TextDatumGetCString(nodeName), WORKER_LENGTH);
+	strlcpy(workerNode->workerRack, TextDatumGetCString(nodeRack), WORKER_LENGTH);
+
+	return workerNode;
+}
+
+
+/*
+ * master_get_next_nodeid allocates and returns a unique nodeId for the node
+ * to be added. This allocation occurs both in shared memory and in write
+ * ahead logs; writing to logs avoids the risk of having nodeId collisions.
+ *
+ * Please note that the caller is still responsible for finalizing node data
+ * and the nodeId with the master node. Further note that this function relies
+ * on an internal sequence created in initdb to generate unique identifiers.
+ *
+ * NB: This can be called by any user; for now we have decided that that's
+ * ok. We might want to restrict this to users part of a specific role or such
+ * at some later point.
+ */
+Datum
+master_get_next_nodeid(PG_FUNCTION_ARGS)
```

Comment:
We don't actually exercise these functions as UDFs, would it make sense to just turn them into regular functions?